hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/blockmanagement/TestBlockManagerSafeMode.java

```diff
@@ -230,6 +232,50 @@ public void testCheckSafeMode8() throws Exception {
     assertEquals(BMSafeModeStatus.OFF, getSafeModeStatus());
   }
 
+  @Test(timeout = 20000)
+  public void testCheckSafeMode9() throws Exception {
+    Configuration conf = new HdfsConfiguration();
+    try {
+      conf.setLong(DFS_NAMENODE_SAFEMODE_RECHECK_INTERVAL_KEY, 3000);
+      bm = spy(new BlockManager(fsn, false, conf));
+      doReturn(true).when(bm).isGenStampInFuture(any(Block.class));
+      dn = spy(bm.getDatanodeManager());
+      Whitebox.setInternalState(bm, "datanodeManager", dn);
+      // the datanode threshold is always met
+      when(dn.getNumLiveDataNodes()).thenReturn(DATANODE_NUM);
+      bmSafeMode = new BlockManagerSafeMode(bm, fsn, false, conf);
+      bmSafeMode.activate(BLOCK_TOTAL);
+      Whitebox.setInternalState(bmSafeMode, "extension", Integer.MAX_VALUE);
+      setSafeModeStatus(BMSafeModeStatus.PENDING_THRESHOLD);
+      setBlockSafe(BLOCK_THRESHOLD);
+      bmSafeMode.checkSafeMode();
+
+      assertTrue(bmSafeMode.isInSafeMode());
+      assertEquals(BMSafeModeStatus.EXTENSION, getSafeModeStatus());
+
+      GenericTestUtils.waitFor(new Supplier<Boolean>() {
+        @Override
+        public Boolean get() {
+          Whitebox.setInternalState(bmSafeMode, "extension", 0);
+          return getSafeModeStatus() != BMSafeModeStatus.EXTENSION;
+        }
+      }, EXTENSION / 10, EXTENSION * 10);
+
+      assertFalse(bmSafeMode.isInSafeMode());
+      assertEquals(BMSafeModeStatus.OFF, getSafeModeStatus());
+    } finally {
+      conf.setLong(DFS_NAMENODE_SAFEMODE_RECHECK_INTERVAL_KEY,
+          DFS_NAMENODE_SAFEMODE_RECHECK_INTERVAL_DEFAULT);
+      bm = spy(new BlockManager(fsn, false, conf));
+      doReturn(true).when(bm).isGenStampInFuture(any(Block.class));
+      dn = spy(bm.getDatanodeManager());
+      Whitebox.setInternalState(bm, "datanodeManager", dn);
+      // the datanode threshold is always met
+      when(dn.getNumLiveDataNodes()).thenReturn(DATANODE_NUM);
+      bmSafeMode = new BlockManagerSafeMode(bm, fsn, false, conf);
+    }
```

Comment:
The tests have too many warnings due to deprecation. See if you can get rid of them, If not, just do a assert on the log output, That it contains ``Using <value> as SafeModeMonitor Interval``
``GenericTestUtils`` has a ``LogCapturer`` you can use that.